### ISSUE TITLE
YoloV8 Edge TPU Export Input Unclear

### ISSUE THREAD

**Issue by team955** [ultralytics/ultralytics#9010]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

(Note: this model was trained and exported using YoloV8, so there should be no issues that result from transferring data from other models)

I have been trying to create a YoloV8 model working on the Google Coral Edge TPU. I have gotten things mostly working, and I have a tflite that works on CPU. It is claimed to be "int8 quantized" but it still takes in float16 values, so I doubt that it as actually int8 quantized, which is a bit confusing (more evidence will come below)

Here is the python code I wrote for the "int8 quantized" model.

[CPU_tflite.txt](https://github.com/ultralytics/ultralytics/files/14629428/CPU_tflite.txt)

As you can see, the "int8 quantized" model actually takes in float16, and it seems to process them fine, accurately predicting what I want it to predict. I tried to do the same with the edgetpu export, but I was not able to get any working values. I suspect it is because the input the the previously "int8 quantized" model takes in values that are between -1 and 1, which all get set to zero when using the truly int8 quantized edgetpu export. Below is a screenshot of my input data to the CPU "int8 quantized" model, as well as the inference time and then the returned results.

![Screenshot from 2024-03-17 16-57-07](https://github.com/ultralytics/ultralytics/assets/556881/b10fd0ff-9706-43c4-8743-b6e34a2efec4)

(You can tell it is running on the CPU due to the 1.5 second inference time, as well as how much CPU it uses)

So, I think when the YOLOv8 tries to quantize this model to work on the edgetpu, it is not doing anything with the inputs, causing the model to still be tuned for values between -1 and 1, while the only values supported are -1, 0, and 1, causing weird behavior to happen. This is what I see when I try to run the new model.

![Screenshot from 2024-03-17 17-01-56](https://github.com/ultralytics/ultralytics/assets/556881/1a294288-a19a-413b-ae1c-6688223ef0b7)

(Running on the TPU, from the 0.06 second inference time)

As expected, all the input values are zero, and the model has some very weird outputs, exclusively outputting 0s, -128s, and occasionally a -127. I also tried removing the divide by the input std, and got this:

![Screenshot from 2024-03-17 17-05-19](https://github.com/ultralytics/ultralytics/assets/556881/ae91cefd-cb52-4857-95aa-88ebc19bf5ba)

Again, it still only outputs -128, -127, and 0. I cannot seems to find any information about this, and any help would be appreciated.

What I would really want to know:

What inputs does this "edgetpu" export take in, and what kind of outputs does it take out?
What is up with the "int8 quantized" model that does not really seem int8 quantized?
If the edgetpu export takes in the same values as the "int8 quantized" one does, then the inputs would be zero once actually quantized. Is there any way to "scale up" the whole entire model so the inputs now become -127 to 128?

**Comment by team955** [comment]
I doubt it is of any significance, but just in case it is important, I am running an Ubuntu 22.04 on a [Beelink Mini S Mini PC](https://www.bee-link.com/beelink-mini-s-n5095-mini-pc) with a Google Coral USB Accelerator. I have the [unofficial libedgetpu for python3.10](https://github.com/feranick/libedgetpu/releases), as Google seems to have discontinued the Coral support, and all the Coral tests run fine. If there is any other information needed, please let me know

**Comment by glenn-jocher** [comment]
@team955 thank you for reaching out with your detailed observation! It sounds like you're making great strides with YOLOv8 and the Coral Edge TPU. Let's address your questions:

1. **Input and Output for Edge TPU Export:** The Edge TPU expects models to be fully quantized, including input and output layers. For best performance, your input should indeed be INT8 values ranging between -128 and 127. When exporting your model for the Edge TPU, it is crucial that both your model and your input data match this expectation. 

2. **INT8 Quantized Models Accepting Float16:** The behavior you're observing may stem from the preprocessing steps. TensorFlow and TFLite, in their conversion or inference processes, might be automatically handling the data conversion for you. This is not uncommon and is sometimes seen to ease the transition from training to deployment environments. However, for Edge TPU, strict adherence to quantization rules is essential.

3. **Model Scaling:** To address the scaling issue (inputs being reduced to 0 when properly quantized), you should implement pre-processing that scales the floating-point input values to the entire INT8 range before quantization. Below is a simple Python code snippet on how you might approach scaling your input data:

    ```python
    def scale_input_to_int8(input_data):
        # Assuming input_data is your input in float32
        # Scale input data to the INT8 range (-128, 127)
        scaled_input = np.clip(input_data * 127.0, -128, 127).astype(np.int8)
        return scaled_input
    ```

    Make sure to apply a similar inverse scaling operation to your model's output if necessary, to interpret the predictions correctly.

Regarding the "int8 quantized" model that doesn't strictly seem int8 quantized, it's possible that during the export or conversion process, the quantization parameters (scale and zero-point) ensure that the effective numerical range and resolution accommodate float16 inputs. This technique maintains backward compatibility and supports varied inference scenarios.

For your setup on Ubuntu 22.04 with the Beelink Mini S and the Google Coral USB Accelerator, it looks like everything is in place. The unofficial libedgetpu for Python 3.10 should suffice for running Edge TPU models, provided that the Coral tests are passing as you mentioned.

If you encounter further issues or need more specific guidance, feel free to share additional details or reach out again. We're here to help! 🚀

**Comment by team955** [comment]
Hello, thank you for your help. As shown above, I have tried putting the model through with inputs -127 to 128, and they still end up with strange inputs and outputs. So I was wondering exactly how should I preprocess the image, and it would be very nice to see some examples on how to scale the data properly, or if the model even works (it does not seem to), since whatever I do I always get -127, 0, or -128 (which is equal to 128). Thank you for clairifying that the int8 quantized model takes in data that has not been completely processed yet, but my question still holds on how exactly to preprocess the data so that the edgetpu exported model can interpret it.

I have tried looking through the source code, but I am a bit confused. I will continue to look, but any help would be appreciated. Thanks!

**Comment by team955** [comment]
Got it! It was just how I was processing the output, as well as adding the function above! Thanks for the help!

Here is my code, for anyone that wants help

import numpy as np
import cv2
import os
import pathlib
import time
import tflite_runtime.interpreter as tflite

input_mean = 127.5
input_std = 127.5

img_width = 640
img_height = 640

script_dir = pathlib.Path('').parent.absolute()
model_file = os.path.join(script_dir, 'v8n_50/best_full_integer_quant_edgetpu.tflite')
label_file = os.path.join(script_dir, 'v8n_50/labels.txt')
vid = cv2.VideoCapture(0)

interpreter = tflite.Interpreter(model_file, experimental_delegates=[tflite.load_delegate('libedgetpu.so.1', options={"device": "usb"})])
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

scale, zero_point = input_details[0]["quantization"]

while(True):
	(_, im) = vid.read()
	rgb = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)
	resize = cv2.resize(rgb, (640, 640), interpolation = cv2.INTER_LINEAR)
	inputData = np.expand_dims(resize, axis = 0)
	inputData = (np.float32(inputData) - input_mean) / input_std
	inputData = np.clip(inputData * 127.0, -128, 127).astype(np.int8)
	
	print(inputData)

	interpreter.set_tensor(input_details[0]['index'], inputData)
	
	start = time.perf_counter()
	interpreter.invoke()
	inference_time = time.perf_counter() - start
	
	print(inference_time)
	
	output = interpreter.get_tensor(output_details[0]["index"])
	
	scale, zero_point = output_details[0]["quantization"]
	output = (output.astype(np.float32) - zero_point) * scale
	
	output[:, [0, 2]] *= 640
	output[:, [1, 3]] *= 640

	print(output)
        
	boxes = []
	scores = []
	class_ids = []
	
	for pred in output:
		pred = np.transpose(pred)
		for box in pred:
			x, y, w, h = box[:4]
			x1 = x - w / 2
			y1 = y - h / 2
			boxes.append([x1, y1, w, h])
			scores.append(box[4])

	indices = cv2.dnn.NMSBoxes(boxes, scores, 0.3, 0.4)

	print(indices)
	
	output_image = resize
        
	for i in indices:
		# Get the box, score, and class ID corresponding to the index
		box = boxes[i]
		score = scores[i]
		if score > 0:
			print(score)
			output_image = cv2.rectangle(resize, (int(box[0]), int(box[1])), (int(box[0]+box[2]), int(box[3]+box[1])), np.random.uniform(0, 255, 3), 2)
	
	cv2.imshow("Output", output_image)
	
	cv2.waitKey(0)

**Comment by team955** [comment]
(forgot to close I'm a dummy)

**Comment by glenn-jocher** [comment]
@team955 no worries at all, it happens! If you've got everything resolved, that's great news. Should you run into any more questions or need further assistance down the road, don't hesitate to reach out. Happy coding! 😊🚀

**Comment by glenn-jocher** [comment]
Hello! It sounds like you're encountering a segmentation fault when trying to run your model at higher resolutions, which can sometimes occur due to hardware limitations or model conversion issues. Here are a couple of suggestions to help troubleshoot and potentially resolve the issue:

1. **Check Memory Availability**: Higher resolutions like 640x640 consume significantly more memory. Ensure your system and the Edge TPU USB accelerator have enough available memory to handle the larger input size.

2. **Model Conversion Check**: Verify the model has been properly converted and optimized for the Edge TPU at the desired resolution. Sometimes, models need specific optimization flags or adjustments during the conversion process for larger sizes.

3. **Test Incrementally**: If possible, try intermediate sizes (like 320x320) to pinpoint if there is a specific resolution threshold where the issue starts. This can help identify if it's a gradual resource issue or a specific fault at higher resolutions.

Here’s a simple modification to your existing code to add exception handling, which might help catch any errors during interpreter creation:

```python
try:
    self.interpreter = make_interpreter(str(model_path))
except Exception as e:
    print(f"Failed to create interpreter: {e}")
```

Adding this might provide more details on what's going wrong during the interpreter instantiation. Let me know how it goes! 😊

**Comment by UsamaRehman001** [comment]
its not raising any exception either. its only return the segmentation fault. thats it

**Comment by glenn-jocher** [comment]
Hello! 😊 It looks like the segmentation fault is indeed challenging, especially since the model behaves differently at various sizes. The issue with the empty list in the 'quantization parameters' of the `yolov8n_int8.tflite` model might be the root cause of why only certain sizes are functional. 

From your information, it looks like the successful model at 192x192 has proper quantization parameters, which are essential for the Edge TPU's performance. The lack of quantization params in the 640x640 model suggests a potential issue in how the quantization was applied during the conversion.

A quick suggestion: Double-check the quantization process for the 640x640 model. Ensure that the quantization stats (scale and zero points) are being correctly calculated and applied. Sometimes manually specifying these during the conversion process can help, particularly if the automated tools aren't picking them up properly. 

Here's a snippet on how you could apply quantization manually if you're using TensorFlow's converter:

```python
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8  # Force input to be int8
converter.inference_output_type = tf.int8  # Force output to be int8
tflite_model_quant = converter.convert()
```

Also, using TensorFlow's `--enable_numeric_verify` flag during compilation might give more insights into what's going wrong. Keep us updated, and feel free to share any further details! 👍

**Comment by UsamaRehman001** [comment]
in above suggestion saved_model_dir should be yolov8n.pt? i wanna convert yolov8n.pt for image size 640. please give suggestion according to that.

`# Export a YOLOv8n PyTorch model to TFLite Edge TPU format
yolo export model=yolov8n.pt format=edgetpu  # creates 'yolov8n_full_integer_quant_edgetpu.tflite'

# Run inference with the exported model
yolo predict model=yolov8n_full_integer_quant_edgetpu.tflite source='https://ultralytics.com/images/bus.jpg'`

i used above code to test also. but it also give segmentation fault for the "yolov8n_full_integer_quant_edgetpu.tflite" file

**Comment by glenn-jocher** [comment]
Hey there! It looks like you're facing some challenges with segmentation faults when working with larger image sizes on the Edge TPU using YOLOv8. The logs you provided hint that certain operations didn't map properly to the Edge TPU at higher resolutions, which might be causing the issues you're seeing, especially with the segmentation faults.

The `yolov8n.pt` file is indeed a PyTorch model file, and you're in the right direction for exporting it for the Edge TPU. However, be aware that increasing the image size significantly impacts whether operations can be successfully mapped to the Edge TPU due to the memory constraints and operational compatibility.

For now, a good step would be to initially ensure that all your layers and operations are fully compatible with the Edge TPU at smaller sizes (like your working 192 model). Gradually scale up the image size and monitor the compiler's log to catch any operations that might be causing the fault. Sometimes even alternative layer designs or slight changes in parameters might help in getting more layers mapped efficiently.

Unfortunately, if certain layers or configurations inherently don't map well at larger sizes due to TPU limitations or the model's complexity, it might become necessary to optimize or alter parts of the network for better compatibility or even stick to the max feasible resolution.

Fingers crossed, and I hope this pushes you a bit forward! Keep experimenting! 👍🤖

**Comment by bogdannedelcu** [comment]
@UsamaRehman001 It is not an issue with Yolov8, the problem is with conversion from pytorch to tflite, when generating an operation which is not suported by the EdgeTPU. Open the file you just converted in netron.app and look at the first not-converted operation. In my case there is a TRANSPOSE which probably does not fit the EdgeTPU.

The cause of the problem is described here and is related to the way channels are represented in pytorch versus tensorflow. TFLite needs a transpose before each conv2d, because the conv2d need the channels in the last position.

https://github.com/google-ai-edge/ai-edge-torch/issues/179#issuecomment-2363060213
